KhanhNT/linux-2020-7-2.md

```diff
@@ -0,0 +1,280 @@
+# Linux
+## Understand shell environment variables
+Cách chạy 1 file shell ( đuôi .sh): 
+
+- bash hello.sh
+- sh hello.sh
+
+- ./hello.sh
+
+Biến trong shell:
+	
++ Không thể sử dụng các ký tự khác như !, *, hoặc - , bởi vì các ký tự này có ý nghĩa đặc biệt đối với shell.
+
++ Biến hệ thống: là chữ in hoa
+	
++ Biến do người dùng đặt: chữ thường và phải bắt đầu bằng kí tự, không được bằng số
++ Không có giá trị khởi tạo thì bằng NULL
+	
+Variable_name=variable value
+VD: VAR1=”khanh”
+       VAR2=100
+
+Để truy cập các giá trị được giữ trong một biến, đặt trước tên biến đó với ký hiệu ( $)
+
+VD: NAME = “Khanh”
+       echo $NAME
+	
+Không có dấu cách 2 bên toán tử = khi gán giá trị cho biến
+
+	
+Shell cung cấp một cách để đánh dấu các biến như read-only bằng cách sử dụng lệnh readonly. Sau khi một biến được đánh dấu read-only, giá trị của nó không thể thay đổi.
+
+Xóa 1 biến trong linux: Xóa một biến nói cho shell để gỡ bỏ biến đó từ danh sách các biến mà nó theo dõi.
+
+NAME = “Khanh”
+
+unset NAME 
+
+       echo $NAME
+
+note: Không thể dùng unset xóa biến readonly
+	
+Xem toàn bộ biến môi trường: env
+	
+Các kí tự toán học
+expr toántử 1 toánhạng toántử 2 
+Các dấu ngoặc:
+ 
+-	Tất cả các kí tự trong dấu ngoặc kép đều không có giá trị tính toán 
+
+-	Dấu nháy ngược: yêu cầu thực hiện lệnh
+
+Cấu trúc điều khiển trong shell script:
+	
+## GETTING HELP
+ Sử dụng lệnh man để tra cứu
+
+Các option của 1 số lệnh cơ bản: 
+
+###	Lệnh ls: 
+
++ ls -l: Hiển thị thư mục và các file có trong thư mục hiện tại với nhiều thông tin hơn như các quyền read, write, excute
+
++ ls -a: Hiển thị các file có thư mục ẩn
+
++ ls *.txt: hiển thị các file có đuôi là txt
+
+###	Lệnh cd:
+cd / : di chuyển đến thư mục gốc của user hiện tại
+cd ..: lùi về thu mục trước đó
+
+###	Lệnh cat: 
+cat filename.txt: hiển thị trong file
+cat >filename.txt: xóa trong file và thêm dữ liệu mới vào file
+
+###	Lệnh rm: 
+rm : xóa thư mục trống
+
+rm -rf: xóa thư mục là các file, thư mục con của nó.
+
+## Using stream, Redirection, and Pipes
+•	Stream 
+
+Standard Input (stdin): có file desciptor là 0, trong hầu hết các trường hợp data đến từ keyboard của người dùng.
+
+Standard Output (stdout): có file descriptor là 1, thể hiện trên màn hình dạng text-mode hoặc GUI.
+
+Standard Error (stderr): có file descriptor là 2, thể hiện các thông tin quan trọng như lỗi ra màn hình như stdout, ta có thể tách biệt stderr ra 1 file riêng so với stdout để theo dõi.
+
+•	Redirecting:
```

Comment:
@khanhnt99 
đọc thêm  https://www.cyberciti.biz/faq/how-to-redirect-output-and-errors-to-devnull/